### PR TITLE
Use new Authorize.net APIs for payment processing.

### PR DESCRIPTION
**Pull request by charleswaddell**
Used by https://github.com/silverorange/course-host/pull/1225

### CODE REVIEW

**Comment by charleswaddell** [comment]
https://app.clubhouse.io/silverorangeinc/story/8051/authorize-net-sandbox-not-working

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 1]
```diff
@@ -418,29 +412,111 @@ protected function getIPAddress()
...
+				} else if (is_array($messages) && count($messages) > 0){
+					$message = $messages[0];
+
+					$text = sprintf(
+						'%s: %s',
+						$message->getCode(),
+						$message->getText()
+					);
+				}
+			} else if (is_array($messages) && count($messages) > 0){
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
			} elseif (is_array($messages) && count($messages) > 0){
```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 2]
```diff
@@ -418,29 +412,111 @@ protected function getIPAddress()
...
+			$messages = $response->getMessages()->getMessage();
+			$tresponse = $response->getTransactionResponse();
+			if ($tresponse instanceof AnetAPI\TransactionResponseType) {
+				$errors = $tresponse->getErrors();
+				if (is_array($errors) && count($errors) > 0) {
+					$error = $errors[0];
+
+					$code = $error->getErrorCode();
+					$text = $error->getErrorText();
+				} else if (is_array($messages) && count($messages) > 0){
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
				} elseif (is_array($messages) && count($messages) > 0){
```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 3]
```diff
@@ -418,29 +412,111 @@ protected function getIPAddress()
...
+		return $auth;
+	}
+
+	// }}}
+	// {{{ protected function hasError()
+
+	protected function hasError($response)
+	{
+		if ($response instanceof AnetAPI\AnetApiResponseType) {
+			if ($response->getMessages()->getResultCode() != "Ok") {
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
			if ($response->getMessages()->getResultCode() !== 'Ok') {
```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 4]
```diff
@@ -339,66 +307,92 @@ protected function setRequestCardFields(
...
-			$item->id,
-			$this->truncateField($item->product_title, 31),
-			$this->truncateField($item->description, 255),
-			$item->quantity,
-			$this->formatNumber($item->price),
-			false
-		);
+	protected function getLineItem(StoreOrderItem $item)
+	{
+		$line_item = new AnetAPI\LineItemType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$line_item = new AnetAPI\LineItemType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 5]
```diff
@@ -418,29 +412,111 @@ protected function getIPAddress()
...
+	// {{{ protected function getTax()
 
-	protected function truncateField($content, $maxlength)
+	protected function getTax(StoreOrder $order)
 	{
-		$content = SwatString::condense($content, $maxlength - 4, ' ...');
-		$content = str_replace('  •  ', ' - ', $content);
-		$content = html_entity_decode($content, ENT_QUOTES, 'ISO-8859-1');
-		return $content;
+		$amount = new AnetAPI\ExtendedAmountType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$amount = new AnetAPI\ExtendedAmountType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 6]
```diff
@@ -418,29 +412,111 @@ protected function getIPAddress()
...
+
+		return $amount;
+	}
+
+	// }}}
+	// {{{ protected function getMerchantAuthentication()
+
+	protected function getMerchantAuthentication()
+	{
+		$auth = new AnetAPI\MerchantAuthenticationType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$auth = new AnetAPI\MerchantAuthenticationType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 7]
```diff
@@ -418,29 +412,111 @@ protected function getIPAddress()
...
 
-	/**
-	 * @param float $value
-	 *
-	 * @return string formatted .
-	 */
-	protected function formatNumber($value)
+	protected function getShipping(StoreOrder $order)
+	{
+		$amount = new AnetAPI\ExtendedAmountType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$amount = new AnetAPI\ExtendedAmountType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 8]
```diff
@@ -339,66 +307,92 @@ protected function setRequestCardFields(
...
+
+		return $ord;
+	}
+
+	// }}}
+	// {{{ protected function getCustomer()
+
+	protected function getCustomer(StoreOrder $order)
+	{
+		$customer = new AnetAPI\CustomerDataType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$customer = new AnetAPI\CustomerDataType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 9]
```diff
@@ -339,66 +307,92 @@ protected function setRequestCardFields(
...
+		$customer->setEmail($order->email);
+
+		if ($order->account !== null && $order->account->id !== null) {
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		if ($order->account instanceof SiteAccount && $order->account->id !== null) {
```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 10]
```diff
@@ -339,66 +307,92 @@ protected function setRequestCardFields(
...
 
-	protected function addRequestLineItems(
-		AuthorizeNetAIM $request,
-		StoreOrder $order
-	) {
-		foreach ($order->items as $item) {
-			$this->addRequestLineItem($request, $item);
+	protected function getOrder(StoreOrder $order)
+	{
+		$ord = new AnetAPI\OrderType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$ord = new AnetAPI\OrderType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 11]
```diff
@@ -339,66 +307,92 @@ protected function setRequestCardFields(
...
 
-	protected function setRequestAddressFields(
-		AuthorizeNetAIM $request,
-		StoreOrderAddress $address
-	) {
-		$request->first_name = $address->first_name;
-		$request->last_name  = $address->last_name;
+	protected function getBillTo(StoreOrderAddress $address)
+	{
+		$addr = new AnetAPI\CustomerAddressType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]:
```suggestion
		$addr = new AnetAPI\CustomerAddressType();

```

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 12]
```diff
@@ -312,21 +280,21 @@ protected function getOrderDescription(StoreOrder $order)
...
-	protected function setRequestCardFields(
-		AuthorizeNetAIM $request,
+	protected function getPayment(
 		StoreOrder $order,
 		$card_number,
 		$card_verification_value = null
 	) {
-		$request->card_num = $card_number;
-		$request->card_code = $card_verification_value;
+		$creditCard = new AnetAPI\CreditCardType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]: `creditCard` => `credit_card`

File: Store/StoreAuthorizeNetPaymentProvider.php [hunk 13]
```diff
@@ -134,156 +137,121 @@ public function pay(
...
-			$card_number,
-			$card_verification_value
-		);
-
-		// Order fields
-		$request->invoice_num = $this->getInvoiceNumber($order);
-		$request->description = $this->truncateField(
-			$this->getOrderDescription($order),
-			255
+		$requestType = new AnetAPI\TransactionRequestType();
```
Comment (gauthierm) [inline Store/StoreAuthorizeNetPaymentProvider.php]: `requestType` => `request_type`

**Pull request merged**